contracts/concordium-governance-committee-election/src/lib.rs

```diff
@@ -194,33 +209,175 @@ impl ElectionConfig {
+pub type RegisterGuardianPublicKeyParameter = Vec<u8>;
+
+/// Entrypoint for registering a public key for the guardian corresponding to
+/// the sender address.
+#[receive(
+    contract = "election",
+    name = "registerGuardianPublicKey",
+    parameter = "RegisterGuardianPublicKeyParameter",
+    error = "Error",
+    mutable
+)]
+fn register_guardian_public_key(ctx: &ReceiveContext, host: &mut Host<State>) -> Result<(), Error> {
+    let mut guardian_state = validate_guardian_context(ctx, host)?;
+    ensure!(guardian_state.public_key.is_none(), Error::DuplicateEntry);
+
+    let parameter: RegisterGuardianPublicKeyParameter = ctx.parameter_cursor().get()?;
+    guardian_state.public_key = Some(parameter);
+    Ok(())
+}
+
+/// The parameter expected by the [`register_guardian_final_key`] entrypoint.
+pub type RegisterGuardianEncryptedShareParameter = Vec<u8>;
+
+/// Entrypoint for registering an encryption share for the guardian
+/// corresponding to the sender address.
+#[receive(
+    contract = "election",
+    name = "registerGuardianEncryptedShare",
+    parameter = "RegisterGuardianEncryptedShareParameter",
+    error = "Error",
+    mutable
+)]
+fn register_guardian_final_key(ctx: &ReceiveContext, host: &mut Host<State>) -> Result<(), Error> {
+    let mut guardian_state = validate_guardian_context(ctx, host)?;
+    ensure!(
+        guardian_state.encrypted_share.is_none(),
+        Error::DuplicateEntry
+    );
+
+    let parameter: RegisterGuardianEncryptedShareParameter = ctx.parameter_cursor().get()?;
+    guardian_state.encrypted_share = Some(parameter);
+    Ok(())
+}
+
+/// Entrypoint for filing a comlaint if the cumulative state of guardians cannot
+/// be verified.
+#[receive(
+    contract = "election",
+    name = "registerGuardianStatus",
+    parameter = "GuardianStatus",
+    error = "Error",
+    mutable
+)]
+fn register_guardian_status(ctx: &ReceiveContext, host: &mut Host<State>) -> Result<(), Error> {
+    let mut guardian_state = validate_guardian_context(ctx, host)?;
+    ensure!(guardian_state.status.is_none(), Error::DuplicateEntry);
+
+    let status: GuardianStatus = ctx.parameter_cursor().get()?;
+    guardian_state.status = Some(status);
+    Ok(())
+}
+
+/// The cumulative state of all guardians
+pub type GuardiansState = Vec<(AccountAddress, GuardianState)>;
+
+/// View the cumulative state of all guardians. This is useful for guardians to
+/// validate the key registrations of other guardians, and construct final keys
+/// from registered pre-keys.
+#[receive(
+    contract = "election",
+    name = "viewGuardiansState",
+    return_value = "GuardiansState"
+)]
+fn view_guardians_state(
+    _ctx: &ReceiveContext,
+    host: &Host<State>,
+) -> ReceiveResult<GuardiansState> {
+    let guardians_state = &host.state.guardians;
+    let guardians_state: Vec<_> = guardians_state
+        .iter()
+        .map(|(address, guardian_state)| (*address, guardian_state.clone()))
+        .collect();
```

Comment:
It always hurts me when we do such needless cloning. We could be more fancy by having GuardianState contain references, since this value just gets serialized at the end.

But in reality it does not matter much since this is just a view function.

So I probably should not have written this comment :P

Reply:
I'm always happy to receive suggestions as to how to avoid needless clones, so keep the coming! 😄

Reply:
So I tried changing `GuardiansState` to hold references to `GuardianState`'s, but couldn't get it to work due some ownership error. Basically tried to call `defef` instead of `clone` on the `guardian_state` inside the map function, but it complained that I tried to return a reference to a value owned by the function.

I'll leave it as is for now (although I would love to see how the error would be resolved).